app/javascript/src/AccessTokens/components/ExpirationDatePicker.tsx

```diff
@@ -0,0 +1,176 @@
+  { id: '7', label: '7 days', period: 7 },
+  { id: '30', label: '30 days', period: 30 },
+  { id: '60', label: '60 days', period: 60 },
+  { id: '90', label: '90 days', period: 90 },
+  { id: 'custom', label: 'Custom...', period: 0 },
+  { id: 'no-exp', label: 'No expiration', period: 0 }
+]
+
+const dayMs = 60 * 60 * 24 * 1000
+
+interface Props {
+  id: string;
+  label: string | null;
+  tzOffset?: number;
+}
+
+const ExpirationDatePicker: FunctionComponent<Props> = ({ id, label, tzOffset }) => {
+  const today: Date = new Date()
+  const tomorrow: Date = new Date(today)
+  tomorrow.setDate(today.getDate() + 1)
+  const [dropdownSelectedItem, setDropdownSelectedItem] = useState(collection[0])
+  const [calendarPickedDate, setCalendarPickedDate] = useState(tomorrow)
+  const fieldName = `human_${id}`
+  const fieldLabel = label ?? 'Expires in'
+
+  const dropdownDate = useMemo(() => {
+    if (dropdownSelectedItem.period === 0) return null
+
+    return new Date(today.getTime() + dropdownSelectedItem.period * dayMs)
+  }, [dropdownSelectedItem])
+
+  const selectedDate = useMemo(() => {
+    let value = null
+
+    if (dropdownDate) {
+      value = dropdownDate
+    } else if (dropdownSelectedItem.id === 'custom' ) {
+      value = calendarPickedDate
+    }
+
+    return value
+  }, [dropdownDate, dropdownSelectedItem, calendarPickedDate])
+
+  const formattedDateValue = useMemo(() => {
+    if (!selectedDate) return
+
+    const formatter = Intl.DateTimeFormat('en-US', {
+      month: 'long', day: 'numeric', year: 'numeric', hour: 'numeric', minute: 'numeric', hour12: false
+    })
+
+    return formatter.format(selectedDate)
+  }, [selectedDate])
+
+  const fieldHint = useMemo(() => {
+    if (!formattedDateValue) return
+
+    return `The token will expire on ${formattedDateValue}`
+  }, [formattedDateValue])
+
+  const inputDateValue = useMemo(() => {
+    if (!selectedDate) return ''
+
+    return selectedDate.toISOString()
+  }, [selectedDate])
+
+  const tzMismatch = useMemo(() => {
+    if (tzOffset === undefined) return
+
+    // Timezone offset in the same format as ActiveSupport
+    const jsTzOffset = new Date().getTimezoneOffset() * -60
+
+    return jsTzOffset !== tzOffset
+  }, [tzOffset])
+
+  const labelIcon = useMemo(() => {
+    if (!tzMismatch) return
+
+    return (
+      <Popover
+        bodyContent={(
+          <p>
+            Your local time zone differs from the provider default.
+            The token will expire at the time you selected in your local time zone.
+          </p>
+        )}
+        headerContent={(
+          <span>Time zone mismatch</span>
+        )}
+      >
+        <button
+          aria-describedby="form-group-label-info"
+          aria-label="Time zone mismatch warning"
+          className="pf-c-form__group-label-help"
+          type="button"
+        >
+          <OutlinedQuestionCircleIcon noVerticalAlign />
+        </button>
+      </Popover>
+    )
+  }, [tzMismatch])
```

Comment:
Not sure what's the rationale here, but using `useMemo` is generally a bad idea. Memoization in React is nothing like Rails' (or other proper languages) and in the end it makes it less performant.

Reply:
I use it to declare a variable as reactive, e.g.:

```typescript
const selectedDate = useMemo(() => {
    ...
  }, [dropdownDate, dropdownSelectedItem, calendarPickedDate])
```

`selectedDate` must be re-evaluated whenever `dropdownDate`, `dropdownSelectedItem` or `calendarPickedDate` change.

Is this not the right way to do it?

Reply:
`useMemo` is for memoization of very heavy computations. What you want you usually do in the onChange methods as a callback or simply declaring the variables in the body of the component. Every time there is a change, the component will re-render and every variable re-evaluated.

For instance `inputDateValue` depends directly on `selectedDate`, so you simply do:
```
  const inputDateValue = selectedDate?.toISOString() // -> string | undefined
```

Reply:
I refactored here: https://github.com/3scale/porta/pull/3943/commits/4d8a47c74c14a0d858782b8b78b816ef13691064